### PR TITLE
Llm eval script

### PR DESCRIPTION
**Pull request by KeremKurban**

### CODE REVIEW

File: tests/tools/test_validate_tool_call.py [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertFalse(result)
+
+    def test_no_required_tools_but_optional(self):
+        result, message = validate_tool(
+            required_tools=[],
+            actual_tool_calls=["tool1", "tool2"],
+            optional_tools=["tool1", "tool2"],
```
Comment (WonderPG) [inline tests/tools/test_validate_tool_call.py:98]: Does/should the order of the optional tools matter ? Real case exemple:
get features from a certain morpho with tortuosity < 1:
- `required: [resolve]`
- `optional: [get_morpho, morpho_feature, kg_morpho_feature]`
Here `get_morpho` should happen before `morpho_feature`, but `kg_morpho_feature` can happen anytime in theory.

File: tests/tools/test_validate_tool_call.py [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+    def test_only_optional_tools_called(self):
+        result, message = validate_tool(
+            required_tools=[],
+            actual_tool_calls=["tool2"],
+            optional_tools=["tool2"],
+            forbidden_tools=[],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
```
Comment (WonderPG) [inline tests/tools/test_validate_tool_call.py:111]: Isn't this the same test as the one above ?

File: tests/tools/test_validate_tool_call.py [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+            ],
+            optional_tools=[],
+            forbidden_tools=[],
+        )
+        self.assertTrue(result)
+        self.assertEqual(message, "All required tools called correctly")
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Comment (WonderPG) [inline tests/tools/test_validate_tool_call.py:260]: Could you add a small test for the function that calls the API (mocking it) ?

**Review by WonderPG — COMMENTED** [review]
Thank you for this ! We might find edge cases when running it, but it's fine.

File: src/neuroagent/scripts/validate_tool_calls.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    Returns
+    -------
+        tuple: A tuple containing a boolean and a string message. The boolean is True if the
+               validation is successful, otherwise False. The string message provides details
+               about the validation result.
+    """
+    # Check for forbidden tools
+    for tool in actual_tool_calls:
+        if tool in forbidden_tools:
+            return False, f"Forbidden tool called: {tool}"
```
[the remark this answers is not in the document]
Comment (KeremKurban) [reply src/neuroagent/scripts/validate_tool_calls.py:42]: oh lala, indeed

File: src/neuroagent/scripts/validate_tool_calls.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+        # Check if the response is successful
+        if response.status_code == 200:
+            # Parse the response
+            steps = response.json().get("steps", [])
+            called_tool_names = [step.get("tool_name", None) for step in steps]
+            expected_tool_names = [
+                tool_call.get("tool_name", None) for tool_call in expected_tool_calls
+            ]
+            match, _ = validate_tool(
```
[the remark this answers is not in the document]
Comment (KeremKurban) [reply src/neuroagent/scripts/validate_tool_calls.py:111]: what do you mean by reason ?

File: tests/tools/test_validate_tool_call.py (hunk 1 above)
Comment (KeremKurban) [reply tests/tools/test_validate_tool_call.py:98]: so either it can fetch morph and calculate by itself or it can fetch from kg directly. so maybe both options should be ok

File: tests/tools/test_validate_tool_call.py (hunk 2 above)
Comment (KeremKurban) [reply tests/tools/test_validate_tool_call.py:111]: indeed, deleting other

File: tests/tools/test_validate_tool_call.py (hunk 3 above)
Comment (KeremKurban) [reply tests/tools/test_validate_tool_call.py:259]: adding it for async case. maybe we dont need sync case then

**Comment by KeremKurban** [comment]
I added async version and added max 10 concurrent requests along with feedbacks in both sync and async. if its ok , i can delete the sync one

File: src/neuroagent/scripts/avalidate_tool_calls.py [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    async with semaphore:
+        prompt = query["prompt"]
+        expected_tool_calls = query["expected_tools"]
+        optional_tools = query["optional_tools"]
+        forbidden_tools = query["forbidden_tools"]
+
+        logging.info(f"Testing prompt: {prompt}")
+
+        # Send a request to the API
+        async with session.post(
```
Comment (WonderPG) [inline src/neuroagent/scripts/avalidate_tool_calls.py:51]: This doesn't need to be a context manager.

File: src/neuroagent/scripts/avalidate_tool_calls.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+            f"{base_url}/qa/run",
+            headers={"Content-Type": "application/json"},
+            json={
+                "query": prompt,
+                "messages": [{"role": "user", "content": prompt}],
```
Comment (WonderPG) [inline src/neuroagent/scripts/avalidate_tool_calls.py:56]: The `messages` part is not needed for our API.

File: src/neuroagent/scripts/avalidate_tool_calls.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+    format="%(asctime)s - %(levelname)s - %(message)s",  # Define the log message format
+)
+
+
+async def fetch_tool_call(
+    session: aiohttp.ClientSession,
+    query: Dict[str, Any],
+    base_url: str,
+    semaphore: asyncio.Semaphore,
+) -> Dict[str, Any]:
```
Comment (WonderPG) [inline src/neuroagent/scripts/avalidate_tool_calls.py:23]: No need to import Dict from typing, lower case `dict` is available

File: src/neuroagent/scripts/avalidate_tool_calls.py [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+    the results.
+
+    The function is designed to be the entry point when the script is run
+    directly from the command line.
+    """
+    parser = argparse.ArgumentParser(
+        description="Run tool call tests and save results."
+    )
+    parser.add_argument(
+        "--base",
```
Comment (WonderPG) [inline src/neuroagent/scripts/avalidate_tool_calls.py:202]: I would rather name it `base_url` to avoid confusion

File: src/neuroagent/scripts/avalidate_tool_calls.py [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+    parser.add_argument(
+        "--output",
+        type=str,
+        default="tool_call_evaluation.csv",
+        help="Output CSV file for results",
+    )
+    args = parser.parse_args()
+
+    # Construct the full base URL
+    base_url = f"http://{args.base}:{args.port}"
```
Comment (WonderPG) [inline src/neuroagent/scripts/avalidate_tool_calls.py:228]: This would fail if we point to aws for instance. I suggest we let the user include the protocol in the `base_url`

File: src/neuroagent/scripts/avalidate_tool_calls.py (hunk 6 above)
Comment (KeremKurban) [reply src/neuroagent/scripts/avalidate_tool_calls.py:51]: I resaerched the pros and cons of this vs await approach. Mainly , this one is better for resource management while await is better for flexibility.

File: tests/tools/test_validate_tool_call.py (hunk 1 above)
Comment (KeremKurban) [reply tests/tools/test_validate_tool_call.py]: Lets keep this in mind in future

**Comment by KeremKurban** [comment]
@WonderPG can you approve ?

**Review by WonderPG — APPROVED** [review]
